cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java

```diff
@@ -197,21 +198,28 @@ public void run() {
           } else {
             AnomalyNotificationResult notificationResult = null;
             // Call the anomaly notifier to see if a fix is desired.
-            if (anomaly instanceof BrokerFailures) {
-              BrokerFailures brokerFailures = (BrokerFailures) anomaly;
-              _anomalyDetectorState.addAnomaly(AnomalyType.BROKER_FAILURE, brokerFailures);
-              notificationResult = _anomalyNotifier.onBrokerFailure(brokerFailures);
-              _brokerFailureRate.mark();
-            } else if (anomaly instanceof GoalViolations) {
-              GoalViolations goalViolations = (GoalViolations) anomaly;
-              _anomalyDetectorState.addAnomaly(AnomalyType.GOAL_VIOLATION, goalViolations);
-              notificationResult = _anomalyNotifier.onGoalViolation(goalViolations);
-              _goalViolationRate.mark();
-            } else if (anomaly instanceof KafkaMetricAnomaly) {
-              KafkaMetricAnomaly metricAnomaly = (KafkaMetricAnomaly) anomaly;
-              _anomalyDetectorState.addAnomaly(AnomalyType.METRIC_ANOMALY, metricAnomaly);
-              notificationResult = _anomalyNotifier.onMetricAnomaly(metricAnomaly);
-              _metricAnomalyRate.mark();
+            AnomalyType anomalyType = getAnomalyType(anomaly);
+            switch (anomalyType) {
+              case GOAL_VIOLATION:
+                GoalViolations goalViolations = (GoalViolations) anomaly;
+                _anomalyDetectorState.addAnomalyDetection(anomalyType, goalViolations);
+                notificationResult = _anomalyNotifier.onGoalViolation(goalViolations);
+                _goalViolationRate.mark();
+                break;
+              case BROKER_FAILURE:
+                BrokerFailures brokerFailures = (BrokerFailures) anomaly;
+                _anomalyDetectorState.addAnomalyDetection(anomalyType, brokerFailures);
```

Comment:
for the same broker failure, we may call this multiple time( the first time get a `CHECK_WITH_DELAY`) and each time it will create a new `AnomalyState` and reset the `_detectionMs` to current time(which is not the true anomaly detection time)

Reply:
TL;DR -- This is intentional.
For the broker failure detection, we enforce a grace period -- i.e. before taking an action on broker failures, we intentionally wait for a certain period time. So the final mitigation action of an initial broker failure should be `CHECK_WITH_DELAY`, and the following one should be `FIX_STARTED` if the self-healing is enabled for broker failures. Ideally, we would like to preserve this transparency to indicate when wast the first time the failure was detected, and when was the second time the same failure was identified.

Reply:
I see, so `CHECK_WITH_DELAY` will actually create a new anomaly.